Use correct top origin when checking firstPartyForCookies with site isolation enabled

achristensen07 opened the pull request:
#### e8a014af82e87656aefc463f85ffb0e1e386b5bf
<pre>
Use correct top origin when checking firstPartyForCookies with site isolation enabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=272373">https://bugs.webkit.org/show_bug.cgi?id=272373</a>
<a href="https://rdar.apple.com/126114826">rdar://126114826</a>

Reviewed by Charlie Wolfe and Sihui Liu.

There were two issues here:

1. Document::topOrigin was using topDocument() to get the SecurityOrigin, and with site isolation
enabled the top document isn&apos;t necessarily the main frame&apos;s Document.  Instead, use a strategy
like we do with mainFrameURL where the Page has the main frame&apos;s origin in all processes, whether
the main frame is remote or not.

2. In WebProcessPool::processForNavigation when navigating the main frame, use the RegistrableDomain
from the navigation instead of the PageLoadState, which may not have the activeURL yet.

With these 2 changes, we stop getting failures in NetworkProcess::allowsFirstPartyForCookies in 2 tests
when run with --site-isolation.  The tests change from crashes to regular time-outs.

* LayoutTests/platform/mac-site-isolation/TestExpectations:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::topOrigin const):
* Source/WebCore/dom/Document.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setMainFrameURL):
(WebCore::Page::mainFrameOrigin const):
* Source/WebCore/page/Page.h:
* Source/WebCore/page/SecurityOrigin.cpp:
(WebCore::SecurityOrigin::opaqueOrigin):
* Source/WebCore/page/SecurityOrigin.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::processForNavigation):

Canonical link: <a href="https://commits.webkit.org/277331@main">https://commits.webkit.org/277331@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b4d167ca403e89aac9ae3f2f89c7a1e5d2b0e256

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47349 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26529 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/50000 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50033 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43398 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/49656 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/32041 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23989 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38553 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47930 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/24069 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40804 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19873 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/21508 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41970 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5393 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/43703 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/42372 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51908 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22380 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/18719 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/45856 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23654 "Built successfully") | | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/44886 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24440 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6667 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23377 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -977,6 +977,18 @@ Ref<SecurityOrigin> Document::protectedTopOrigin() const
+            return mainFrameDocument->securityOrigin();
+    }
+    if (RefPtr page = this->page())
+        return page->mainFrameOrigin();
```
In existing implementation, we run check on security policy (`SecurityContext::securityOrigin()`) for `Document::topOrigin()`, I don't see the security check in this patch. Do we run it when setting main frame URL for the subframe process?

achristensen07 replied:
What checks are done in SecurityContext::securityOrigin?  We either don't need to do them in a remote frame process or we haven't implemented them for site isolation yet.

szewai replied:
SecurityContext::securityOrigin is decided by securityOriginPolicy, and securityOriginPolicy can be set via SecurityContext::setSecurityOriginPolicy. It seems in most cases it will be SecurityOrigin of the main document, but can be opaque (e.g. if the iframe is sandboxed).

achristensen07 replied:
I think that goes in the category of things that haven't been implemented in site isolation yet, but this PR is an improvement still.

szewai replied:
~~ok let's add a FIXME about adding check for SecurityOriginPolicy of top document later.~~
hmm actually this seems to introduce behavior change for non-site-isolation case; can we guard it with site isolation flag?

achristensen07 replied:
Great idea!

charliewolfe commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -977,6 +977,18 @@ Ref<SecurityOrigin> Document::protectedTopOrigin() const
+SecurityOrigin& Document::topOrigin() const
+{
+    RefPtr frame = this->frame();
+    if (RefPtr localMainFrame = frame ? dynamicDowncast<LocalFrame>(frame->mainFrame()) : nullptr) {
```
Seems unfortunate that we have to do this. When do we rely on getting the top document security origin when `Page` does not exist?

achristensen07 replied:
We don't, but we'd rather have an opaque origin than crash if it does happen during teardown.

charliewolfe replied:
I was referring to the downcast. It would be nice to remove this to match behavior when the main frame is out-of-process.

achristensen07 replied:
Yeah, there's still work to do here.

achristensen07 committed:
Use correct top origin when checking firstPartyForCookies with site isolation enabled
https://bugs.webkit.org/show_bug.cgi?id=272373
rdar://126114826

Reviewed by Charlie Wolfe and Sihui Liu.

There were two issues here:

1. Document::topOrigin was using topDocument() to get the SecurityOrigin, and with site isolation
enabled the top document isn't necessarily the main frame's Document.  Instead, use a strategy
like we do with mainFrameURL where the Page has the main frame's origin in all processes, whether
the main frame is remote or not.

2. In WebProcessPool::processForNavigation when navigating the main frame, use the RegistrableDomain
from the navigation instead of the PageLoadState, which may not have the activeURL yet.

With these 2 changes, we stop getting failures in NetworkProcess::allowsFirstPartyForCookies in 2 tests
when run with --site-isolation.  The tests change from crashes to regular time-outs.

* LayoutTests/platform/mac-site-isolation/TestExpectations:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::topOrigin const):
* Source/WebCore/dom/Document.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setMainFrameURL):
(WebCore::Page::mainFrameOrigin const):
* Source/WebCore/page/Page.h:
* Source/WebCore/page/SecurityOrigin.cpp:
(WebCore::SecurityOrigin::opaqueOrigin):
* Source/WebCore/page/SecurityOrigin.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::processForNavigation):

Canonical link: https://commits.webkit.org/277331@main
